### PR TITLE
Dsl2 refactor

### PR DESCRIPTION
**Pull request by MaxUlysse**
## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
  - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
  - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/sarek/tree/master/.github/CONTRIBUTING.md)
  - [ ] If necessary, also make a PR on the nf-core/sarek _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint .`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: modules/nf-core/software/bwamem2_index.nf
Comment (FriederikeHanssen) [inline modules/nf-core/software/bwamem2_index.nf:14]: The tool version here and in the actual mapping module is not the same. should be 2.1 as well

**Comment by abhi18av** [comment]
Hi @MaxUlysse  👋 

What's the plan to add `DeepVariant` to the DSL2 version?

This current PR looks pretty extensive, any plans on merging this one to `dsl2` branch soon? I can use that one (after the merge) as a foundation.

**Comment by MaxUlysse** [comment]
Hi @abhi18av
You got the main idea.
I do need a couple of more stuff to be fully up to date, but it should arrive soon.
Current plan is to merge this PR to `dsl2` by the end of the week, that way we will have caught up with the current DSL2 best practices.
And from there, start to work on `deepvariant` and other issues.

**Comment by abhi18av** [comment]
Ah, cool 👍 

Also, I wanted to ask whether you'd like to include the [stub](https://www.nextflow.io/docs/edge/process.html#stub) feature as well while we are at it?

**Comment by MaxUlysse** [comment]
That's a point I need to check with the @nf-core/core team.
But I do like the idea

**Comment by MaxUlysse** [comment]
OK, so I think current state is good enough for me for now.
and from that we can really start working on finishing the port.
There's still some linting issue to figure out, but that'll be done when `tools` is updated and ready for `DSL2`

**Comment by MaxUlysse** [comment]
There will also be more modules update, but I really think that we can merge that before, and then go crazy with the modules and (sub)workflows.
I think there's room for improvement in the current sarek workflow (in the workflow folder).
I'll try to put more code in `lib/` so that we can have dedicated workflow depending on what we want to do.
-> thinking about regular sarek, sarek with minimal genome, sarek with sentieon, etc...

File: modules/nf-core/software/bwa/mem/functions.nf
Comment (FriederikeHanssen) [inline modules/nf-core/software/bwa/mem/functions.nf:1]: These function files are the same for each module, right? Wouldn't it make sense to only have them once then? Sorry, if there was a discussion on this already and I missed it 🙈

Comment (MaxUlysse) [reply modules/nf-core/software/bwa/mem/functions.nf:1]: They are the same for now.
But I think the point of having them separate is that when you update a module, you update its `functions.nf` companion at the same time, so you can modules that are not all as up to date with the nf-core/modules repo.
But yeah, it's a discussion we had somewhere, and I think that's the consensus for now.
I assume one improvement could be to have a tagged version of this file, so that we use it only once per repo/version and not one per module

Comment (abhi18av) [reply modules/nf-core/software/bwa/mem/functions.nf:1]: Yeah, Harshil mentioned that these would be unnecessary once the newer version of `nf-core/tools` is released and can take care of deps common to all `nf-core/modules` such as the `functions.nf`. 

In the meantime, we toil with repetition 🚴

File: .github/workflows/ci.yml
Comment (FriederikeHanssen) [inline .github/workflows/ci.yml:5]: is this now running on all branches? or none?

File: .github/workflows/ci.yml
Comment (FriederikeHanssen) [inline .github/workflows/ci.yml:22]: any reason to not add singularity and conda? To keep the tests short? I am just thinking that given we pull from 3 different sources now, this could catch incorrect urls and such quickly

File: README.md
Comment (FriederikeHanssen) [inline README.md:26]: It comes with 'Docker'... is this still up-to date or should we add the other container engines, since we don't have a monolithic Docker file anymore?

File: README.md
Comment (FriederikeHanssen) [inline README.md:40]: Should be MarkDuplicates due to param changes 
```suggestion
* Mark Duplicates (`GATK MarkDuplicates`)
```

File: README.md
Comment (FriederikeHanssen) [inline README.md:131]:
```suggestion
For further information or help, don't hesitate to get in touch on the [Slack `#sarek` channel](https://nfcore.slack.com/channels/sarek) (you can join with [this invite](https://nf-co.re/join/slack)), or contact us: [Maxime Garcia](mailto:maxime.garcia@scilifelab.se?subject=[GitHub]%20nf-core/sarek), [Szilvester Juhos](mailto:szilveszter.juhos@scilifelab.se?subject=[GitHub]%20nf-core/sarek), [Friederike Hanssen](mailto:friederike.hanssen@qbic.uni-tuebingen.de?subject=[GitHub]%20nf-core/sarek), [Gisela Gabernet](mailto:gisela.gabernet@qbic.uni-tuebingen.de?subject=[GitHub]%20nf-core/sarek)
```

File: modules/local/concat_vcf.nf
Comment (FriederikeHanssen) [inline modules/local/concat_vcf.nf:15]: Does this also have a build number? ` --hd3b49d5_0` I mean?

File: modules/nf-core/software/bwa/mem/main.nf
Comment (FriederikeHanssen) [inline modules/nf-core/software/bwa/mem/main.nf:14]: Why not use samtoll 1.11? as in the other steps? Ah I guess, it is from the modules repo and not a local one, right?

File: modules/nf-core/software/functions.nf
Comment (FriederikeHanssen) [inline modules/nf-core/software/functions.nf:46]: Since the functions file is now within each tool/module folder, is this one still needed?

**Review by FriederikeHanssen — APPROVED** [review]
So beautiful 😍 I just had some minor comments. One thing, I am wondering though is, whether it would make sense to tag the conda versions more specifically as you did in the latest environment.yml: `bwa-mem2=2.0=he513fc3_1` and so on

File: .github/workflows/ci.yml
Comment (FriederikeHanssen) [reply .github/workflows/ci.yml:22]: random spaces accidentally added somewhere in between has caused me a few issues here and there 😄

Comment (MaxUlysse) [reply .github/workflows/ci.yml:22]: Plan is to do that, but not implemented yet

File: README.md
Comment (MaxUlysse) [reply README.md:26]: Good catch, we might need to update the TEMPLATE

Comment (MaxUlysse) [reply README.md:26]: -> https://github.com/nf-core/tools/pull/850

File: modules/nf-core/software/bwa/mem/main.nf
Comment (MaxUlysse) [reply modules/nf-core/software/bwa/mem/main.nf:14]: Because I did not check if there was a mulled container for it actually...
I just used the modules from nf-core/modules when there was one.
Was not planning to update tools in this PR.
I'm saving that for the future PRs

File: modules/local/concat_vcf.nf
Comment (MaxUlysse) [reply modules/local/concat_vcf.nf:15]: Good point, I'll look into it

File: README.md
Comment (FriederikeHanssen) [reply README.md:26]: that was fast 🚒

**Pull request merged**
